airbyte-cdk/python/airbyte_cdk/sources/declarative/parsers/model_to_component_factory.py

```diff
@@ -0,0 +1,58 @@
+#
+# Copyright (c) 2022 Airbyte, Inc., all rights reserved.
+#
+
+from __future__ import annotations
+
+from typing import Callable, List, Literal, Mapping, Type, Union
+
+from airbyte_cdk.sources.declarative.checks import CheckStream
+from airbyte_cdk.sources.declarative.models.declarative_component_schema import CheckStream as CheckStreamModel
+from airbyte_cdk.sources.declarative.models.declarative_component_schema import DeclarativeStream as DeclarativeStreamModel
+from airbyte_cdk.sources.declarative.models.declarative_component_schema import Spec as SpecModel
+from airbyte_cdk.sources.declarative.spec import Spec
+from airbyte_cdk.sources.declarative.types import Config
+from pydantic import BaseModel
+
+ComponentDefinition: Union[Literal, Mapping, List]
+
+
+def create_check_stream(model: CheckStreamModel, config: Config):
+    return CheckStream(model.stream_names, options={})
+
+
+def create_declarative_stream(model: DeclarativeStreamModel, config: Config):
+    # todo this is just a stub temporarily, but should actually return DeclarativeStream but that requires building the whole object
+    return {}
+
+
+def create_spec(model: SpecModel, config: Config):
+    return Spec(connection_specification=model.connection_specification, documentation_url=model.documentation_url, options={})
+
+
+class ModelToComponentFactory:
+    def create_component(self, model_type: Type[BaseModel], component_definition: ComponentDefinition, config: Config) -> type:
+        component_type = component_definition.get("type")
+        if component_definition.get("type") != model_type.__name__:
+            raise ValueError(f"Expected manifest component of type {model_type}, but received {component_type} instead")
```

Comment:
want to make sure I understand what could cause this error... is this for the case where the parameters match a different type?
eg
```
authenticator:
  type: "BasicHttpAuthenticator"
  api_token: "{{ config['api_key'] }}"
```
where BasicHttpAuthenticator expects a username, not an api_key

Reply:
Not quite. So this is a pretty niche case because the `create_component()` will only serve as the entrypoint for a few different invocations by ManifestDeclarativeSource (CheckStream, DeclarativeStream, and Spec components). But if somehow a schema had:
```
check:
  type: SimpleRetriever
  stream_names: ["applications"]
```

then we would want to fail this. This should be caught by any form of schema validation we do, so I'm also okay w/ just getting rid of this extra check too

Reply:
If we keep it in there should probably be a test for it. But looks like `ValidationError` will be thrown before we end up hitting the codepath that would give `ValueError`, in which case I'd be in favor of removing this check.

Reply:
Yep good suggestion, I'll add some basic tests for this flow, but a majority of the tests will be construction related in the follow on PR.

Reply:
i added the test that covers this codepath